Analysis/source/CalcMulticlassDNN.cxx

```diff
@@ -0,0 +1,376 @@
+        model->input.clear();
+        for (const auto& featureName : model->getFeatureNames()) {
+          model->input.setValue(featureName, features.get(featureName));
+        }
+
+        model->run(features.getEventId());
+      }
+
+      // fill the output tree
+      outTree->Fill();
+
+      // print progress
+      if (i == 0 || (i + 1) % args_.progress() == 0 || i == nEntries - 1) {
+        std::cout << "processed entry " << (i + 1) << std::endl;
+      }
+    }
+
+    // write the output file
+    std::cout << "writing output tree" << std::endl;
+    outputFile_->Write();
+
+    // close files and finish
+    outputFile_->Close();
+    inputFile_->Close();
+    std::cout << "done" << std::endl;
+  }
+
+  private:
+  CalcMulticlassDNNArguments args_;
+  std::shared_ptr<TFile> inputFile_;
+  std::shared_ptr<TFile> outputFile_;
+};
+
+FeatureProvider::FeatureProvider(Period period, Channel channel, TTree* inTree)
+    : period_(period)
+    , channel_(channel) {
+  // define names of variables to read
+  std::vector<std::string> boolNames = { "pass_VBF_trigger" };
+  std::vector<std::string> intNames = { "b1_valid", "b2_valid", "VBF1_valid", "VBF2_valid" };
+  std::vector<std::string> ulong64Names = { "evt" };
+  std::vector<std::string> floatNames = { "b1_pt", "b1_eta", "b1_phi", "b1_m", "b1_DeepFlavour",
+    "b1_DeepFlavour_CvsB", "b1_DeepFlavour_CvsL", "b1_HHbtag", "b2_pt", "b2_eta", "b2_phi", "b2_m",
+    "b2_DeepFlavour", "b2_DeepFlavour_CvsB", "b2_DeepFlavour_CvsL", "b2_HHbtag", "VBF1_pt",
+    "VBF1_eta", "VBF1_phi", "VBF1_m", "VBF1_DeepFlavour", "VBF1_DeepFlavour_CvsB",
+    "VBF1_DeepFlavour_CvsL", "VBF1_HHbtag", "VBF2_pt", "VBF2_eta", "VBF2_phi", "VBF2_m",
+    "VBF2_DeepFlavour", "VBF2_DeepFlavour_CvsB", "VBF2_DeepFlavour_CvsL", "VBF2_HHbtag", "tau1_pt",
+    "tau1_eta", "tau1_phi", "tau1_m", "tau2_pt", "tau2_eta", "tau2_phi", "tau2_m", "MET_pt",
+    "MET_phi" };
+
+  // register them in input maps and set branch addresses
+  for (const auto& name : boolNames) {
+    boolInputs_.emplace(name, 0.);
+    inTree->SetBranchAddress(name.c_str(), &boolInputs_.at(name));
+  }
+  for (const auto& name : intNames) {
+    intInputs_.emplace(name, 0.);
+    inTree->SetBranchAddress(name.c_str(), &intInputs_.at(name));
+  }
+  for (const auto& name : ulong64Names) {
+    ulong64Inputs_.emplace(name, 0);
+    inTree->SetBranchAddress(name.c_str(), &ulong64Inputs_.at(name));
+  }
+  for (const auto& name : floatNames) {
+    floatInputs_.emplace(name, 0.);
+    inTree->SetBranchAddress(name.c_str(), &floatInputs_.at(name));
+  }
+}
+
+void FeatureProvider::calculate() {
+  // check if objects are set
+  bool b1Set = intInputs_.at("b1_valid") == 1;
+  bool b2Set = intInputs_.at("b2_valid") == 1;
+  bool vbfj1Set = intInputs_.at("VBF1_valid") == 1;
+  bool vbfj2Set = intInputs_.at("VBF2_valid") == 1;
+  bool lep1Set = floatInputs_.at("tau1_pt") > 0;
+  bool lep2Set = floatInputs_.at("tau2_pt") > 0;
```

Comment:
`lep1Set` and `lep2Set` should always be true. If you encountered some event for which it is not the case, please, let us know, because it would indicate that there is a bug in our code...

Reply:
This was actually just a precaution. I will remove it to be sensible to errors which should not occur.